Add a WK2 input method test for event ordering

rniwa opened the pull request:
#### 8c464456429f7fcc106ad9bc0ec58b8385fdd809
<pre>
Add a WK2 input method test for event ordering
<a href="https://bugs.webkit.org/show_bug.cgi?id=295825">https://bugs.webkit.org/show_bug.cgi?id=295825</a>

Reviewed by Wenson Hsieh.

Add a test for 297270@main.

* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(-[MockTextInputContextAction initWithMarkedText:selectedRange:replacementRange:]):
(-[MockTextInputContext handleEventByInputMethod:completionHandler:]):
(-[TestWebViewWithMockTextInputContext mockInputContext]):
(-[TestWebViewWithMockTextInputContext inputContext]):
(-[TestWebViewWithMockTextInputContext _web_superInputContext]):
(TestWebKitAPI::TEST(WKWebViewMacEditingTests, KeyDownFiresBeforeCompositionEvent)):

Canonical link: <a href="https://commits.webkit.org/297302@main">https://commits.webkit.org/297302@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6bc35c4bbf8ae7d7534ad15683e623822c1c7074

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111259 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117290 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113221 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39507 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84561 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100168 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65007 "run-api-tests-without-change (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24579 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18311 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61110 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120407 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38308 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28462 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38684 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93313 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38410 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16185 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34293 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38197 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37862 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41195 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -193,6 +273,40 @@ - (NSRange)range
+        [[MockTextInputContextAction alloc] initWithMarkedText:@"n" selectedRange:NSRange(0, 1) replacementRange:NSRange(NSNotFound, 0)],
+        [[MockTextInputContextAction alloc] initWithMarkedText:@"ni" selectedRange:NSRange(0, 2) replacementRange:NSRange(NSNotFound, 0)],
+        [[MockTextInputContextAction alloc] initWithMarkedText:@"\u4F60" selectedRange:NSRange(0, 1) replacementRange:NSRange(NSNotFound, 0)],
+    ].mutableCopy;
```
Looks like these are memory leaks — might need to make these autoreleased or use `adoptNS(…).get()`, and then `_actions` would be a `RetainPtr<NSMutableArray>` in `MockTextInputContext`.

(The `-mutableCopy` also needs to be adopted to avoid leaking)

rniwa replied:
oh, right. I'll add them to autorelease pool.

whsieh commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
```diff
@@ -193,6 +273,40 @@ - (NSRange)range
+{
+    __block bool isDone = false;
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
I _think_ we prefer `RetainPtr` over `auto` per smart pointer adoption guidelines?

```suggestion
    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```

rniwa replied:
Fixed.

rniwa committed:
Add a WK2 input method test for event ordering
https://bugs.webkit.org/show_bug.cgi?id=295825

Reviewed by Wenson Hsieh.

Add a test for 297270@main.

* Tools/TestWebKitAPI/Tests/mac/WKWebViewMacEditingTests.mm:
(-[MockTextInputContextAction initWithMarkedText:selectedRange:replacementRange:]):
(-[MockTextInputContext handleEventByInputMethod:completionHandler:]):
(-[TestWebViewWithMockTextInputContext mockInputContext]):
(-[TestWebViewWithMockTextInputContext inputContext]):
(-[TestWebViewWithMockTextInputContext _web_superInputContext]):
(TestWebKitAPI::TEST(WKWebViewMacEditingTests, KeyDownFiresBeforeCompositionEvent)):

Canonical link: https://commits.webkit.org/297302@main
